packages/core/src/network.js

```diff
@@ -8,16 +8,50 @@ const ALLOWED_STATUSES = [200, 201, 301, 302, 304, 307, 308];
 const ALLOWED_RESOURCES = ['Document', 'Stylesheet', 'Image', 'Media', 'Font', 'Other'];
 const ABORTED_MESSAGE = 'Request was aborted by browser';
 
+// DefaultMap, which returns a default value for an uninitialized key
+// Similar to defaultDict in python
+class DefaultMap extends Map {
+  constructor(getDefaultValue, ...mapConstructorArgs) {
+    super(...mapConstructorArgs);
+
+    if (typeof getDefaultValue !== 'function') {
+      throw new Error('getDefaultValue must be a function');
+    }
+
+    this.getDefaultValue = getDefaultValue;
+  }
+
+  get = (key) => {
+    if (!this.has(key)) {
+      this.set(key, this.getDefaultValue(key));
+    }
+
+    return super.get(key);
+  };
+};
+
+// RequestLifeCycleHandler handles life cycle of a requestId
+// Ideal flow:          requestWillBeSent -> requestPaused -> responseReceived -> loadingFinished / loadingFailed
+// ServiceWorker flow:  requestWillBeSent -> responseReceived -> loadingFinished / loadingFailed
+class RequestLifeCycleHandler {
+  constructor() {
+    this.releaseRequest = null;
+    this.releaseResponse = null;
+    this.checkRequestSent = new Promise((resolve) => (this.releaseRequest = resolve));
+    this.checkResponseSent = new Promise((resolve) => (this.releaseResponse = resolve));
+  }
+}
 // The Interceptor class creates common handlers for dealing with intercepting asset requests
 // for a given page using various devtools protocol events and commands.
 export class Network {
   static TIMEOUT = undefined;
 
   log = logger('core:discovery');
 
+  #requestsLifeCycleHandler = new DefaultMap(() => new RequestLifeCycleHandler());
```

Comment:
when is something deleted from this map ?
also considering we have this, we dont need any state sets right ?

Reply:
> when is something deleted from this map ?

Nothing is ever deleted from `#requestsLifeCycleHandler` since it is only used to manage lifecycles.
We can ideally delete `requestId` once all the locks are released from it's object, ie. it has reached end of lifecycle.
But, I don't see the point of it, as this is a very light map.

> also considering we have this, we dont need any state sets right ?

Need `#pending` to communicate from `requestWillBeSent` to `requestPaused` states.
Can remove `#intercepts` completely.
Need `#requests` as it keeps mapping of all active requests and is what `idle` depends on.
Need `#authentication` as this keeps track of all requests that have already been authenticated.
Need `#aborted` to keep track of aborted requests.

In the current PR, I am planning to only remove `#intercepts` queue as it was used for communicating from `requestPaused` to `requestWillBeSent` but that should never happen now.

Reply:
`as this is a very light map.` do not fully agree with this as we could have thousands of requests in a single page load [ in fact we do for many storybook builds ] [ even though not blocking as it would be still few kbs max ]

`_requestPaused` could `await requestWillBeSent`, so it doesnt need to read from `#pending` as far as i understand. Because we are also storing event in pending with request id, you can just `resolve` with `event` for request will be sent 
`#requests` agree we need it for now at least
`#authentication` could be done by checking if auth promise is already resolved [ but i dont mind using set either ]
`#aborted` is not necessary as you could just set `aborted` on `#request` object [ or in lifecycle like lifecyle.aborted ] it was done the way it is because we could process `aborted` event before `requestWillBeSent` and it wont exist in requests at all but as `loadingFailed` would await on `requestWillBeSent` we can be sure that request already exists in requests